evm/src/cpu/kernel/tests/bignum/mod.rs

```diff
@@ -0,0 +1,394 @@
+    let len = bignum_len(&a).max(bignum_len(&b));
+    let output_len = len * 2;
+    let memory = pad_bignums(&[a, b], len);
+
+    let a_start_loc = 0;
+    let b_start_loc = len;
+    let output_start_loc = 2 * len;
+    let (new_memory, _new_stack) = run_test(
+        "mul_bignum",
+        memory,
+        vec![
+            len.into(),
+            a_start_loc.into(),
+            b_start_loc.into(),
+            output_start_loc.into(),
+        ],
+    )?;
+
+    let output = mem_vec_to_biguint(&new_memory[output_start_loc..output_start_loc + output_len]);
+    assert_eq!(output, expected_output);
+
+    Ok(())
+}
+
+#[test]
+fn test_shr_bignum_all() -> Result<()> {
+    for bit_size in BIT_SIZES_TO_TEST {
+        let input = gen_bignum(bit_size);
+        let output = input.clone() >> 1;
+        test_shr_bignum(input, output)?;
+
+        let input = max_bignum(bit_size);
+        let output = input.clone() >> 1;
+        test_shr_bignum(input, output)?;
+    }
+
+    let inputs = test_data_biguint(TEST_DATA_BIGNUM_INPUTS);
+    let shr_outputs = test_data_biguint(TEST_DATA_SHR_OUTPUTS);
+    for (input, output) in inputs.iter().zip(shr_outputs.iter()) {
+        test_shr_bignum(input.clone(), output.clone())?;
+    }
+
+    Ok(())
+}
+
+#[test]
+fn test_iszero_bignum_all() -> Result<()> {
+    for bit_size in BIT_SIZES_TO_TEST {
+        let input = gen_bignum(bit_size);
+        let output = input.is_zero() as u8;
+        test_iszero_bignum(input, output.into())?;
+
+        let input = max_bignum(bit_size);
+        let output = 0;
+        test_iszero_bignum(input, output.into())?;
+    }
+
+    let inputs = test_data_biguint(TEST_DATA_BIGNUM_INPUTS);
+    let iszero_outputs = test_data_u256(TEST_DATA_ISZERO_OUTPUTS);
+    let mut iszero_outputs_iter = iszero_outputs.iter();
+    for input in inputs {
+        let output = iszero_outputs_iter.next().unwrap();
+        test_iszero_bignum(input.clone(), *output)?;
+    }
+
+    Ok(())
+}
+
+#[test]
+fn test_cmp_bignum_all() -> Result<()> {
+    for bit_size in BIT_SIZES_TO_TEST {
+        let a = gen_bignum(bit_size);
+        let b = gen_bignum(bit_size);
+        let output = match a.cmp(&b) {
+            Ordering::Less => MINUS_ONE,
+            Ordering::Equal => 0.into(),
+            Ordering::Greater => 1.into(),
+        };
+        test_cmp_bignum(a, b, output)?;
+
+        let a = max_bignum(bit_size);
+        let b = max_bignum(bit_size);
+        let output = 0.into();
+        test_cmp_bignum(a, b, output)?;
+    }
+
+    let inputs = test_data_biguint(TEST_DATA_BIGNUM_INPUTS);
+    let cmp_outputs = test_data_u256(TEST_DATA_CMP_OUTPUTS);
+    let mut cmp_outputs_iter = cmp_outputs.iter();
+    for a in inputs.clone() {
+        for b in inputs.clone() {
+            let output = cmp_outputs_iter.next().unwrap();
+            test_cmp_bignum(a.clone(), b.clone(), *output)?;
+        }
+    }
```

Comment:
Here and in the other tests with similar code, I think you can just iterate over slices rather than cloning the whole input.
```suggestion
    for a in &inputs {
        for b in &inputs {
            let output = cmp_outputs_iter.next().unwrap();
            test_cmp_bignum(a.clone(), b.clone(), *output)?;
        }
    }
```